src/app/portlets/content-types/fields/content-type-fields-add-row/content-type-fields-add-row.component.spec.ts

```diff
@@ -0,0 +1,269 @@
+        'contenttypes.content.three_columns': 'Three columns',
+        'contenttypes.content.four_columns': 'Four columns'
+    });
+
+    beforeEach(() => {
+        testHotKeysMock = new TestHotkeysMock();
+
+        DOTTestBed.configureTestingModule({
+            declarations: [
+                ContentTypeFieldsAddRowComponent
+            ],
+            imports: [TooltipModule, BrowserAnimationsModule],
+            providers: [
+                { provide: HotkeysService, useValue: testHotKeysMock },
+                { provide: MessageService, useValue: messageServiceMock}
+            ]
+        });
+
+        fixture = DOTTestBed.createComponent(ContentTypeFieldsAddRowComponent);
+        de = fixture.debugElement;
+        comp = fixture.componentInstance;
+        el = de.nativeElement;
+    });
+
+    it('should render disabled input', () => {
+        disabled = true;
+        comp.disabled = disabled;
+
+        fixture.detectChanges();
+
+        const buttonElement = de.query(By.css('button'));
+
+        expect(comp.disabled).toEqual(buttonElement.nativeElement.disabled);
+    });
+
+    it('should render columns input', () => {
+        const columnSelectionList = de.query(By.css('.dot-add-rows-columns-list__container'));
+        columns = [1, 2, 3];
+
+        comp.columns = columns;
+
+        fixture.detectChanges();
+
+        expect(comp.columns.length).toEqual(columnSelectionList.children.length);
+    });
+
+    it('should render tooltip input data', () => {
+        toolTips = ['contenttypes.content.one_column',
+        'contenttypes.content.four_columns'];
+
+        comp.ngOnInit();
+        comp.toolTips = toolTips;
+
+        fixture.detectChanges();
+
+        const columnSelectionList = de.query(By.css('.dot-add-rows-columns-list__container'));
+        const columnSelectionItems = columnSelectionList.nativeElement.children;
+
+        expect(comp.i18nMessages[toolTips[1]]).toEqual(columnSelectionItems[1].attributes[9].textContent);
+    });
+
+    it('should display the add rows button by default', () => {
+        const addRowContainer = de.query(By.css('.dot-add-rows-button__container'));
+        const buttonElement = de.query(By.css('button'));
+
+        fixture.detectChanges();
+
+        expect(addRowContainer.nativeElement.classList.contains('dot-add-rows__add')).toEqual(true);
+        expect(buttonElement).toBeTruthy();
+    });
+
+    it('should display row selection after click on Add Rows button', () => {
+        fixture.detectChanges();
+        const addRowContainer = de.query(By.css('.dot-add-rows-columns-list__container'));
+
+        const addButton = de.query(By.css('button'));
+        addButton.nativeElement.click();
+        fixture.detectChanges();
+
+        expect(addRowContainer.nativeElement.classList.contains('dot-add-rows__select')).toEqual(true);
+    });
+
+   it('should focus the first column selection after click on Add Rows button', (done) => {
```

Comment:
What's this ```done``` param here for?

Reply:
use fakeAsync and tick instead

Reply:
I used done param to include done() inside the setTimeout function. I'll use fakeAsync and tick.

Reply:
FakeAsync didn't work, I needed to add the setTimeout function in order to fix the focus issue  with the transition effect. https://stackoverflow.com/questions/10955201/how-to-test-a-function-which-has-a-settimeout-with-jasmine  

This is the TS code:
selectColumnState(): void {
        this.rowState = 'select';
        this.bindKeyboardEvents();
        setTimeout(() => { this.setFocus(this.getElementSelected()); }, 400);
    }